jsonmap.go

```diff
@@ -18,32 +21,96 @@ var (
 	nullRawMessage = RawMessage{nullJSONValue}
 )
 
-type ValidationError struct {
-	reason string
-	rpath  []string
+type FlattenedPathError struct {
+	Path string
+	Message string
+}
+
+func (e *FlattenedPathError) String() string {
+	return fmt.Sprintf("%s: %s\n", e.Path, e.Message)
 }
 
-func NewValidationError(reason string, a ...interface{}) *ValidationError {
-	return &ValidationError{
-		reason: fmt.Sprintf(reason, a...),
-		rpath:  make([]string, 0, 2),
+func NewFlattenedPathError(path, message string) *FlattenedPathError {
+	return &FlattenedPathError{
+		Path: path,
+		Message: message,
 	}
 }
 
+type ValidationError struct {
+	NestedErrors []*FlattenedPathError
+}
+
 func (e *ValidationError) Error() string {
-	msg := e.reason
-	for _, seg := range e.rpath {
-		msg = seg + ": " + msg
+	b := strings.Builder{}
+	b.WriteString("Validation Errors: \n")
+	for _, f := range e.NestedErrors {
+		b.WriteString(f.String())
```

Comment:
Do you expect each error to end with a newline? Do you need to incorporate spacing between the errors here?

Also, `strings.Builder` implements the `io.Writer` interface, so you can use things like `fmt.Fprintf(yourStringBuilder, "%s\n", f.String())` or something similar.